### ISSUE TITLE
Scaling of pods is not happening as per the http requests

### ISSUE THREAD

**Issue by Manikiran88888** [kedacore/http-add-on#1005]
### Report

Scaling of pods is not happening according to the requirements

### Expected Behavior

deployment yaml 
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80 
        resources:
            limits:
              memory: "50Mi" 
              cpu: "50m"
```
http scaled object yaml

```yaml
kind: HTTPScaledObject
apiVersion: http.keda.sh/v1alpha1
metadata:
    name: nginx-http-scaledobject
spec:
    hosts: myhost.com
    targetPendingRequests: 1
    scaleTargetRef:
        deployment: nginx
        service: nginx
        port: 80
    replicas:
        min: 1
        max: 3
```
the scaled object is applied on the deployment

![image](https://github.com/kedacore/http-add-on/assets/86592911/10451574-ef0c-4a49-aa8d-dc138f3dcfba)

### Actual Behavior

The pods should scale up as the load on the http requests is generated but the pods are not scaling 

### Steps to Reproduce the Problem

1.
2.
3.

### Logs from KEDA HTTP operator

```
example
```

### HTTP Add-on Version

0.7.0

### Kubernetes Version

1.29

### Platform

None

**Comment by JorTurFer** [comment]
Hello,
Could you share KEDA operator logs, HTTP add-on interceptor logs? What error do you see in the request?

**Comment by Manikiran88888** [comment]
[test.log](https://github.com/kedacore/http-add-on/files/15157236/test.log)

These are the logs of KEDA operator please find it

**Comment by Manikiran88888** [comment]
> Hello, Could you share KEDA operator logs, HTTP add-on interceptor logs? What error do you see in the request?

**Comment by Manikiran88888** [comment]
> > Hello, Could you share KEDA operator logs, HTTP add-on interceptor logs? What error do you see in the request?

**Comment by Manikiran88888** [comment]
> Hello, Could you share KEDA operator logs, HTTP add-on interceptor logs? What error do you see in the request?

[Uploading test.log…]()

**Comment by Manikiran88888** [comment]
[http.log](https://github.com/kedacore/http-add-on/files/15161317/http.log)

Find the http-interceptor logs

**Comment by JorTurFer** [comment]
The interceptor hasn't received any request. Could you double check the connections? If everything looks well, could you share all the related manifests?
- ingress
- external service
- your workload service service 
- HTTPScaledObject

**Comment by JorTurFer** [comment]
> The interceptor hasn't received any request. Could you double check the connections? If everything looks well, could you share all the related manifests?
> 
> * ingress
> * external service
> * your workload service service
> * HTTPScaledObject

@Manikiran88888 , could you share the info please?

**Comment by Manikiran88888** [comment]
I am sharing the zip folder which consists 

- my yaml file for the deployment of a service, called myntra.yaml where you can change the image name which is in public
- the httpscaledobject yaml file which is deploy-keda-httpscaledobject.yaml.
- I have shared the recent logs of the above mentioned yaml files 

[sample-http.zip](https://github.com/kedacore/http-add-on/files/15175448/sample-http.zip)

please check the files and let me know if any thing required

**Comment by JorTurFer** [comment]
I only see 2 request in the interceptor:
```bash
{"level":"info","ts":1714551507.0846822,"logger":"LoggingMiddleware","caller":"middleware/logging.go:73","msg":"127.0.0.1:51490 - - [01/May/2024:08:18:27 +0000] \"GET / HTTP/1.1\" 404 9 \"\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36\""}
{"level":"info","ts":1714551507.1441922,"logger":"LoggingMiddleware","caller":"middleware/logging.go:73","msg":"127.0.0.1:51490 - - [01/May/2024:08:18:27 +0000] \"GET /favicon.ico HTTP/1.1\" 404 9 \"http://127.0.0.1/\" \"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36\""}
```
Both returning 404.

I see that you are calling using `127.0.0.1`. Are you using a local cluster or so for testing this? If yes, you have to set the `Host` header, like a proxy does. The interceptor relies on it to route the request. The host header has to match with any host set in the HTTPScaledObject

**Comment by JorTurFer** [comment]
Assuming that you HTTPScaledObject sets `myhost.com` as a host, you can try this:
```bash
curl -H `Host: myhost.com` http://127.0.0.1/favicon.ico
```
and that should trigger the scaling out and give you the favicon

**Comment by Manikiran88888** [comment]
I have tested it using the locust where I have given the load 
and I have given in this format 

![image](https://github.com/kedacore/http-add-on/assets/86592911/b0a812f2-7bb0-423d-80b7-76842c8086f6)

because my curl -H command is giving me error in the terminal

![image](https://github.com/kedacore/http-add-on/assets/86592911/e94a8a64-7bae-485f-9f53-549226f898b2)

**Comment by JorTurFer** [comment]
Powershell doesn't have a real curl command, that's why you see that error.
I'm not a PW expert, but I think that this should be equivalent:
```powershell
$headers = @{
    "Host" = "myhost.com"
}
$response = Invoke-WebRequest -Uri "http://127.0.0.1/favicon.ico" -Headers $headers
```

You need to pass the `Host` header if you call directly using the IP address

**Comment by JorTurFer** [comment]
The other option can be [editing the host file](https://docs.digitalocean.com/products/paperspace/machines/how-to/edit-windows-hosts-file/) to include a mapping between your host and 127.0.0.1.
Calling directly to  the IP address without any host won't work as the host is required

**Comment by Manikiran88888** [comment]
I want to know one more thing why should I use **(http://127.0.0.1/favicon.ico)** 
I mean why **favicon.ico** 
because I have port forwarded my application to the port 8090 then I have to use **http://127.0.0.1:8090** if I am not wrong 

I am asking out of curiosity I want to know why favicon.ico came here 

Is it okay if I use  **_curl -H `Host: myhost.com` http://127.0.0.1/8090/_**

**Comment by JorTurFer** [comment]
oh, yeah, you should use the port too.
About the path, my example was with favicon because I saw that path in logs, but you can use the path that you want, the important point is that your request do have to contain the **Host** header as you are calling directly via IP address.

**Comment by Manikiran88888** [comment]
I have tried the complete setup in the wsl and the curl command worked 
![image](https://github.com/kedacore/http-add-on/assets/86592911/800180ce-6b46-40ac-95ec-d9e80aec7f7b)

But the pods did not increase. is there any thing that I need to provide to generate the load

**Comment by JorTurFer** [comment]
Do you want to generate the load from your laptop directly?

**Comment by JorTurFer** [comment]
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: load-generator
  namespace: WHERE_YOU_WANT
spec:
  template:
    spec:
      containers:
      - name: apache-ab
        image: ghcr.io/kedacore/tests-apache-ab
        imagePullPolicy: Always
        args:
          - "-n"
          - "20000" #Total requests
          - "-c"
          - "1"  #Concurrent clients
          - "-H"
          - "Host: YOUR_HOST" 
          - "http://YOUR_INGRESS_SERVICE.YOUR_INGRESS_NAMESPACE:YOUR_INGRESS_PORT/"
      restartPolicy: Never
      terminationGracePeriodSeconds: 5
  activeDeadlineSeconds: 600
  backoffLimit: 5
```

**Comment by JorTurFer** [comment]
This will generate the load directly inside the cluster, hitting your ingress and passing over all the components, but you can hit directly the add-on interceptor service if you prefer, as you wish

**Comment by Manikiran88888** [comment]
How can we hit the add-on interceptor service directly any reference ? 

I would like to check it please

**Comment by JorTurFer** [comment]
About how to configure the scaling and metrics/thresholds, you could be interested on this feature that will be released soon as part of v0.8.0: https://github.com/kedacore/http-add-on/pull/961

**Comment by Manikiran88888** [comment]
Hey,
I could see that the v0.8.0 is released is there any documentation to use the metrics in the yaml file or is there any example to have a look on it 

Also one more thing is that How much ever I do I could not scale up the pods could you please have a look and try to do it on your PC because it is just a sample nginx image which I want to test 
and I want a screen show to check how you are scaling up the pods 
[Test-http.zip](https://github.com/kedacore/http-add-on/files/15187033/Test-http.zip)

**Comment by JorTurFer** [comment]
You can see the reference page for the HTTPScaledObject here: https://kedacore.github.io/http-add-on/ref/v0.8.0/http_scaled_object.html

**Comment by JorTurFer** [comment]
I've just deployed your resources on my own cluster, and deployed this job into the cluster:
```yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: load-generator
spec:
  template:
    spec:
      containers:
      - name: apache-ab
        image: ghcr.io/kedacore/tests-apache-ab
        imagePullPolicy: Always
        args:
          - "-n"
          - "2000000" #Total requests
          - "-c"
          - "400"  #Concurrent clients
          - "-H"
          - "Host: myhost.com" 
          - "http://keda-add-ons-http-interceptor-proxy.keda:8080/test"
      restartPolicy: Never
      terminationGracePeriodSeconds: 5
  activeDeadlineSeconds: 600
  backoffLimit: 5
```

It has scaled as expected:
![image](https://github.com/kedacore/http-add-on/assets/36899226/fa4b10fd-8bb3-4461-869f-9dfeff896b2b)

You have to ensure that your requests reach the nginx, as your HTTPScaledObject defines the path `/test`, the requests have to be to that path, if you call to '/' but your definition is for `/test`, the interceptor won't route the request neither scale.

Assuming that all the traffic reach the monitored workload, you can try modifying the HTTPScaledObject to use the new aggregated metrics, just adding this section:
```
spec: 
.....
  requestRate:
    granularity: 1s
    targetValue: 100
    window: 1m
```
This will be more accurate for the scaling, but this feature has been released today as part of v0.8.0

**Comment by Manikiran88888** [comment]
Okay 
I will also try once again and check for the same

**Comment by Manikiran88888** [comment]
Thanks You 
I am able to scale up and down the pods according to the requirements 

But I am unable to use these metrics 
spec: 
.....
  requestRate:
    granularity: 1s
    targetValue: 100
    window: 1m
    
 I am getting this errors
 
![image](https://github.com/kedacore/http-add-on/assets/86592911/e24829e2-4246-41b7-9075-20dd4590c536)

but I have upgraded my repo using the following commands 

helm repo add kedacore https://kedacore.github.io/charts
helm repo update
helm install http-add-on kedacore/keda-add-ons-http

Am I doing it properly or can I clone the http-add-on repo and should I run the upgrade 
Or else could you provide me the command to upgrade to v0.8.0 

It would be great if you provide me this info

**Comment by JorTurFer** [comment]
There was a bug in helm chart, so if you installed the version v0.7.0, you will need to uninstall the release before (no worries, this bug has been fixed in v0.8.0 and you won't need to remove the add-on anymore and you will be able to upgrade on-the-fly)

```
# removing previous versions
helm uninstall keda-add-ons -n keda

# verify that all the resources have been deleted, this should be empty
kubectl get crd | grep httpscaledobjects.http.keda.sh

# install new version
helm repo add kedacore https://kedacore.github.io/charts
helm install keda kedacore/keda
helm upgrade --install keda-add-ons-http kedacore/keda-add-ons-http --version 0.8.0 -n keda --create-namespace
```

**Comment by Manikiran88888** [comment]
I have followed the steps you have mentioned but still I am getting the same status that Check the crd's first 
Is there any command to check the version of HTTP-add-on I am using i.e. to confirm that if iam using v0.8.0

**Comment by JorTurFer** [comment]
After installing the add-on, could you execute this command and share the output? 
`kubectl get crd | grep httpscaledobjects.http.keda.sh`

**Comment by JorTurFer** [comment]
Could you share also the output of this command?
`kubectl get crd | grep httpscaledobjects.http.keda.sh`

**Comment by Manikiran88888** [comment]
![image](https://github.com/kedacore/http-add-on/assets/86592911/fc398c08-cde5-4731-b5cb-87ac98c3ac14)

**Comment by Manikiran88888** [comment]
![image](https://github.com/kedacore/http-add-on/assets/86592911/f0fbfc35-fefd-4cc4-bfb9-50496a4905d3)

**Comment by Manikiran88888** [comment]
![image](https://github.com/kedacore/http-add-on/assets/86592911/6fa5c9db-747a-4168-8152-7d1e31071691)

**Comment by Manikiran88888** [comment]
My httpscaledobjects code is 
kind: HTTPScaledObject
apiVersion: http.keda.sh/v1alpha1
metadata:
    name: test-httpscale
spec:
    hosts:
    - myhost.com
    pathPrefixes:
    - /test
    scaleTargetRef:
        name: nginx-test-app
        kind: Deployment
        apiVersion: apps/v1
        service: nginx-test-app
        port: 80
    replicas:
        min: 0
        max: 10
    scaledownPeriod: 100
    requestRate:
        granularity: 1s
        targetValue: 100
        window: 1m
        
    Should I need to change anything here

**Comment by JorTurFer** [comment]
Yeah, that's not correct, you can see the correct spec here: https://github.com/kedacore/http-add-on/blob/main/docs/ref/v0.8.0/http_scaled_object.md
the correct place. is:
```
spec:
  scalingMetric:
    requestRate:
      granularity: 1s
      targetValue: 100
      window: 1m
```

**Comment by Manikiran88888** [comment]
My Httpobjectscode is 

`kind: HTTPScaledObject
apiVersion: http.keda.sh/v1alpha1
metadata:
    name: test-httpscale
spec:
    hosts:
    - myhost.com
    pathPrefixes:
    - /test
    scaleTargetRef:
        name: nginx-test-app
        kind: Deployment
        apiVersion: apps/v1
        service: nginx-test-app
        port: 80
    replicas:
        min: 0
        max: 10
    scaledownPeriod: 100
    requestRate:
        granularity: 1s
        targetValue: 100
        window: 1m`
        
     Should I need to change anything here

**Comment by Manikiran88888** [comment]
Thanks for your support on the http topic

**Comment by Manikiran88888** [comment]
Hey,
I have a Query which I want some clarity and It is related to the above code 

After applying the scaling if we want to access the webpage in the browser which has to trigger my Http request and should scale up the pods 
what will be the url which I can use so that I can access my app in the browser 
If you consider the same code as above I want to access the nginx app then which url should I need to use to open it in the browser

**Comment by Manikiran88888** [comment]
Instead of generating the load using the code, I want to open the nginx in the browser and I want to see if it is scaling my pods or mot

**Comment by Manikiran88888** [comment]
Hey, 
In my case the pods are scaling When I hit the load generator 
 
![image](https://github.com/kedacore/http-add-on/assets/86592911/a5d64e1c-f859-414c-b420-6126c5e5ead7)

I have forwarded my interceptor proxy 

![image](https://github.com/kedacore/http-add-on/assets/86592911/bd9f6c03-398d-4915-b11c-36ec40f2f278)
 
but when I try to use the curl command and check the response from the nginx server 

![image](https://github.com/kedacore/http-add-on/assets/86592911/b876d492-9b84-4a12-8096-cf45e68fece7)

these is a not found response I am getting but it is scaling the pods 

Is there anything I am missing or Is there any issue could you please support on it

**Comment by JorTurFer** [comment]
Which is the issue? You are forwarding the request correctly. I mean, you open a port to the interceptor service, and you are calling to it and the interceptor is proxying the request to nginx because that message is from nginx, so it's working as expected.

I guess that you want to open your browser and call to localhost port-forwarded port. Am I right? If yes, it won't work unless you set the headers in your requests using a plugin or so. e.g: https://chromewebstore.google.com/detail/modheader-modify-http-hea/idgpnmonknjnojddfkpgkljpfnnfcklj?pli=1

The problem is that the interceptor routes the traffic based on the `Host` header. This header is set automatically when you call using a domain, but is empty using the IP directly, so reverse proxies don't know how to route it

**Comment by Manikiran88888** [comment]
It worked me for nginx server and it ran properly and when I try to deploy my application 

`apiVersion: apps/v1
kind: Deployment
metadata:
  name: orders-poc
spec:
  replicas: 1
  selector:
    matchLabels: &labels
      app: orders-poc
  template:
    metadata:
      labels: *labels
    spec:
      imagePullSecrets: 
      - name: my-testfapi-test
      containers:
        - name: orders-event
        # Change the image name according to the image created
          image: manikiranreddy58/ordspoc-db:1
          resources:
            limits:
              memory: "200Mi"
              cpu: "300m"

---

kind: Service
apiVersion: v1
metadata:
  name: orders-poc
spec:
  selector:
    app: orders-poc
  ports:
    - protocol: TCP
      port: 9000
      targetPort: 9000`

it is not running and the steps I followed the same how I did for nginx 

![image](https://github.com/kedacore/http-add-on/assets/86592911/0126542c-32bf-48fd-b1af-df1af18044db)

**Comment by JorTurFer** [comment]
The message is quite clear, the connection has been refused :smile:. Most probably because your pod doesn't listen the port 9090 despite it's the configured in k8s side. What do you see if you try to port-forward that port or make a request from any other pod within the cluster?

**Comment by Manikiran88888** [comment]
If I try to port forward it, the app is working fine and it is being accessible 
That is why I want to make clear why is it showing connection refused ?

**Comment by Manikiran88888** [comment]
no actually I am running a fastapi service application and it is giving the error

**Comment by Manikiran88888** [comment]
[sample-http.zip](https://github.com/kedacore/http-add-on/files/15274525/sample-http.zip)


This is the sample code of Fastapi which I am trying to run but it is not executing 

I have deployed it and scaled it also forwarded the port of the interceptor  but when I try the curl command it is not responding it is giving bad gateway

**Comment by JorTurFer** [comment]
> it is not running and the steps I followed the same how I did for nginx
> 
> ![image](https://private-user-images.githubusercontent.com/86592911/328888153-0126542c-32bf-48fd-b1af-df1af18044db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzNDMzMTYsIm5iZiI6MTcxNTM0MzAxNiwicGF0aCI6Ii84NjU5MjkxMS8zMjg4ODgxNTMtMDEyNjU0MmMtMzJiZi00OGZkLWIxYWYtZGYxYWYxODA0NGRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTEwVDEyMTAxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkYjkxNTJmZmU1OGVkZWVlZDBkMjAxNmZlNDY0YWJjZjUwNTJkYWMwMDYyNzZhZTg3NGZhY2Y1ZTJiZTdhMWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.i1sEg7b21VnpqPIp606rRnn4qWcopfCDrjj30fHxJcs)

This picture isn't related with this comment https://github.com/kedacore/http-add-on/issues/1005#issuecomment-2104459910

The picture is pointing to port 9000 and your API doesn't expose that port. Please, don't change all the things all the time without clarifying the change because I'm lost and IDK what you are trying now.

**Comment by JorTurFer** [comment]
Could you please clarify which is the current scenario?:
- What are you deploying?
- What are you executing to validate it?
- Which is the exact error that you are seing?
- Which is the exact behaviour that you are facing?

**Comment by Manikiran88888** [comment]
I have just created the application on port 8008 
Just gave a sample fastapi application 

![image](https://github.com/kedacore/http-add-on/assets/86592911/ee39e570-3f91-45f0-8112-c7090b5b9f43)

**Comment by Manikiran88888** [comment]
What are you deploying?

-  I am deploying a fastapi application for which the code is inside the zip file with a folder name test-fastapi 
- inside the test-fastapi  there is a readme file which is uesful to create the docker image of a fastapi application which will run on port 8008 
- once you create the fastapi application you can use that image in the http-deploy file so that it will deploy the fastapi application on your machine 
- After that you can run the http-scale file to apply the httpscaledobject on the fastapi app


What are you executing to validate it?

- I want to scale up and down this application and also get the response of this application using the curl command 
- the scaling is happening  but there is no response from the curl command 

![image](https://github.com/kedacore/http-add-on/assets/86592911/66c83c41-4080-4791-8752-fd7ae2714454)


Which is the exact error that you are seeing?

- I have forwarded the interceptor proxy 
![image](https://github.com/kedacore/http-add-on/assets/86592911/610ba82f-26e1-4aba-a1a6-6cada890bb75)

- The error is that the connection is refused 
![image](https://github.com/kedacore/http-add-on/assets/86592911/8178dc29-0e23-497e-8872-f90780f51712)

Which is the exact behavior that you are facing?

- when you run the curl command   curl -H "Host: myhost.com" localhost:8080/
- it has to give me a response of hello world 
- but it is prompting bad gateway

**Comment by JorTurFer** [comment]
There are 2 failures:
- your Dockerfile is binding the host to 127.0.0.1, which is not the one when you receive traffic from outside the pod (without a port-forward). I've tested your image and I can't reach the server either form a simple pod next to your deployment, only using port forward. You can try out my image instead yours: `jorturfer/fts`
- your workload doesn't define any readinessProbe, so the container is "ready" before the server is up un running. As the interceptor is monitoring the k8s endpoints, and they are ready when the pod is ready (despite your container isn't ready), the interceptor is proxying the request to a not ready container (because it needs up to 30 seconds to be ready, probably because you have limited the CPU to 50m). You must add readinessProbe to your workload.

Basically, if you change your container to this:
```yaml
        - name: nginx-test-app
          image: jorturfer/fts # use your image name here
          ports:
            - containerPort: 8008
          readinessProbe:
            httpGet:
              path: /
              port: 8008
              scheme: HTTP
```

You will see how it works perfectly.

**Comment by Manikiran88888** [comment]
Okay 
I will try it 
Thanks

**Comment by Manikiran88888** [comment]
Good Morning,
If I use the redinessProbe even after 3 minutes I am still getting this 
![image](https://github.com/kedacore/http-add-on/assets/86592911/9a23895d-afd4-4b20-ad7c-befa0820ac3b)

**Comment by Manikiran88888** [comment]
Hey, 
Good Afternoon 
Could you share the sample code which you have used to create your image `jorturfer/fts`
It is a very important reference for the code which we are using

**Comment by Manikiran88888** [comment]
Good Morning, 
Please could you send us the details of the code which you have used to create the image of fastapi 
We want to see how you have created the image of it because to proceed further with KEDA in our project we want the sample code you have used to create the app of fastapi

**Comment by JorTurFer** [comment]
Hello,
Sorry, I was on a business trip with limited access to GH in general.
I just changed the Dockerfile

```Dockerfile
FROM python:3.10.12-slim

WORKDIR /app

COPY . /app/

RUN pip3 install fastapi uvicorn 

CMD [ "uvicorn", "main:app", "--host", "0.0.0.0", "--port", "8008" ]
```

The important change is this: `"--host", "0.0.0.0"`

**Comment by Manikiran88888** [comment]
Thanks A lot
